server.js

```diff
@@ -1,30 +1,164 @@
+    }
+  } catch (err) {
+    res.status(404).json({ err: 'Page not found' });
+  }
+});
+
+app.get('/thoughts/:thoughtId', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const thought = await Thought.findById(thoughtId).exec();
+    if (thought) {
+      res.json(thought);
+    } else {
+      res.status(404).json({ err: 'Could not find the thought in the database.' });
+    }
+  } catch (err) {
+    res.status(400).json({ err: 'Page not found' });
+  }
+});
+
+app.post('/thoughts', async (req, res) => {
+  try {
+    const newThought = await new Thought({ message: req.body.message, user: req.body.user, tag: req.body.tag }).save();
+    res.status(201).json(newThought);
+  } catch (err) {
+    if (err.code === 11000) {
+      res.status(400).json({ message: 'Duplicated value', fields: err.keyValue });
+    }
+    res.status(400).json({ message: 'Could not save the thought to the database.', err });
+  }
+});
+
+app.post('/thoughts/:thoughtId/like', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const updatedHeart = await Thought.findByIdAndUpdate(thoughtId, { $inc: { hearts: 1 } }, { new: true });
+    if (updatedHeart) {
+      res.json(updatedHeart);
+    } else {
+      res.status(404).json({ message: 'Could not find the thought in the database.' });
+    }
+  } catch (err) {
+    res.status(400).json({ message: 'Could not add a like to the thought in the database.', err });
+  }
+});
+
+app.delete('/thoughts/:thoughtId', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const deletedThought = await Thought.findByIdAndDelete(thoughtId);
+    if (deletedThought) {
+      res.json(deletedThought);
+    } else {
+      res.status(404).json({ message: 'Could not find the thought in the database.' });
+    }
+    res.json(deletedThought);
+  } catch (err) {
+    res.status(400).json({ message: 'Could not delete the thought in the database.', err });
+  }
+});
+
+app.patch('/thoughts/:thoughtId', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const updatedMessage = await Thought.findByIdAndUpdate(thoughtId, req.body, { new: true });
+    if (updatedMessage) {
+      res.json(updatedMessage);
+    } else {
+      res.status(404).json({ message: 'Could not find the thought in the database.' });
+    }
+  } catch (err) {
+    res.status(400).json({ message: 'Could not update the thought in the database.', err });
+  }
+});
+
+app.put('/thoughts/:thoughtId', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const updatedMessage = await Thought.findOneAndReplace({ _id: thoughtId }, req.body, { new: true });
+    if (updatedMessage) {
+      res.json(updatedMessage);
+    } else {
+      res.status(404).json({ message: 'Could not find the thought in the database.' });
+    }
+  } catch (err) {
+    res.status(400).json({ message: 'Could not update the thought in the database.', err });
+  }
+});
```

Comment:
Remember to restrict access to these endpoints because not all users should have access to these functionalities 🤓